### ISSUE TITLE
[WCAG 2.2 Draft Feedback] Success Criterion 2.4.12 Focus Appearance (Minimum)

### ISSUE THREAD

**Issue by dshoukry** [w3c/wcag#1450]
"Success Criterion 2.4.12 Focus Appearance (Minimum) (Level AA): For the keyboard focus indicator of each User Interface Component, all of the following are true:

- Minimum area: The focus indication area is greater than or equal to a 1 CSS pixel border of the focused control, or has a thickness of at least 8 CSS pixels along the shortest side of the element.
- Change of contrast: The color change for the focus indication area has a contrast ratio of at least 3:1 with the colors of the unfocused state.
- Adjacent contrast: The focus indication area has a contrast ratio of at least 3:1 against all adjacent colors for the minimum area or greater, or has a thickness of at least 2 CSS pixels.
- Unobscured: The item with focus is not entirely hidden by author-created content." 

We kindly request significant changes to this SC. Can the group share the research, data, or decision making process that influenced the decisions to require 1, 2, or 8 CSS Pixels as the focus indication area, and the contrast requirements for adjacent items? We provided some proposals about: alternatives other than a glow or border to indicate focus, how to improve some of the examples, and explicitly clarifying that "keyboard focus" doesn't mean "browser focus".  

Please find detailed specifics covered in our [2.4.12 Focus Appearance (Minimum) (Level AA) Google Doc](https://docs.google.com/document/d/1e9LJZ9q13InjvWIq3LJwjNPGiPS8wLD8D7PDKL5YRnM/edit#heading=h.1kj3ui9u2hal).

**Comment by alastc** [comment]
I find it difficult to fit replies into the google doc for some of the longer comments, so I'll reply here (with chair hat off).

@dshoukry asked:
> Can the group share the research, data, or decision making process that influenced the decisions to require 1 or 8 CSS Pixels?

The visibility of an indicator is on a scale. We know that a non-contrasting, 1px dotted outline is really hard to see, for everyone. A 3px highly contrasting solid outline is easy to see for just about everyone.

The main factors that go into the visibility, from some testing and advice from an expert in vision, are:

- Surface area of the indicator, bigger is better.
- The contrast of that surface area is compared the un-focused state, more is better.
- The contrast of the surface area compared to it's adjacent colours, more is better.

For example, a 1px black outline around a white button is reasonably visible. A 3px outline is more visible.
The same 1px black outline around a black button is not very visible at all (example below). A 2px outline makes it appear to change size and is reasonably visible.

But where do we draw the line? It is as much about practicality as user-need, if the requirement is too blunt people won't want to meet it.

In general, having an indicator proportional to the size of the control seems to be a good thing. However, the 8px along the shorter edge came about from #1003, where you could have a quite large indicator (in absolute terms) that fails because the control is very thin & wide.

The 1px border and 8px thick along the shortest side are alternative methods of having sufficient surface area. We're working on making that clearer in #1403 

> Can the minimum area requirement be revised to include minimum width and height size? A one-sided minimum bound could be problematic for small and large controls.

It is an area, rather than a requirement for a 1px border. Anything with the same surface area would be sufficient.

> Ex 1: In small components, like a radio button, if the focus indicator is hugging the visible boundary it may be too small for some users with low vision to see.

Are you suggesting that the size should be increased for smaller controls? We already have another issue with the requirement being too great with some smaller controls (#1329), I'm not sure how to address that without adding a lot of complexity.
Also, as that is the current situation (with browser defaults), I don't think there's a negative impact.


> Ex 2: Can larger controls omit having a focus indication tied directly to area dimensions? For example in a document content area or an email content area, the focus indication could become unproportionally large and distracting.

From my experience A 1px line around a large element isn't to obvious, it is easier to miss than a (smaller in surface area) 2px line around a smaller element. The alternative for components like cards (large area with a heading/link, image & paragraph) is to keep the focus around the text-link within the card, rather than the whole card.


> Additionally, would pointers, tooltip text, or alternative focus indicators other than outlines/borders be sufficient?

This is about keyboard focus, so a pointer change wouldn't be applicable. You could use a tooltip style thing as a focus indicator, it should pass the bullets fairly easily.


> A highly contrastive 7x7px carat or arrow placed adjacent to the focused field might be more apparent than a glow or border.

Possibly, but it does depend how big the element is. E.g. a small arrow in a large 'card' component may not be very visible, but if it is next to links in a navigation then it might be. If it meets the bullets it should be ok.

<hr>

> Can the group share the decision-making process and/or research that informed the decision to proceed with a 2 CSS Pixel thickness or 3:1 contrast ratio?

Going back to the example above, a thin dark outline around a dark button is not very visible (the middle one is focused):
![image](https://user-images.githubusercontent.com/216630/96863462-d2242e80-145e-11eb-9988-d4358730beff.png)

However, if you thicken it, the size makes a difference:
![image](https://user-images.githubusercontent.com/216630/96863502-e23c0e00-145e-11eb-8ad8-410985824d24.png)

This is also applicable is the hue is different, the red might look visible but it does not meet the 3:1 contrast ratio (and would be promblematic for someone with red/green colour blindness):
![image](https://user-images.githubusercontent.com/216630/96863693-2b8c5d80-145f-11eb-92c3-7eda9315f543.png)

From a low vision (and contrast) point of view, the last two examples are the same, so where there is not adjacent contrast then the surface-area change needs to be greater.

> A 2 px focus indicator with low contrast could be problematic. According to some of our own research efforts, a 1px focus indicator with strong contrast was very difficult for participants with minor visual disabilities to see or track. Broad lines were good, but bold, contrastive colors were the top choice participants wanted. 

Indeed, but you need to be tracking what it is contrasting against. If you have a look through some of the [early test-cases](https://alastairc.uk/tests/wcag22-examples/focus-more-visible-2.html) it is a combination of the factors above.

> Given the movement for most visual elements to have a 3:1 contrast per SC 1.4.11 Non-text Contrast, shouldn't focus indicators be included, rather than an approved exception?

1.4.11 doesn't cover this user-requirement very well because it only measure adjacent contrast.

<hr>

> Can the keyboard focus definition and guidance expand to cover indicators beyond browser-level keyboard focus?
> 
> A few examples that may be negatively impacted by the forcing visibility of browser-based keyboard focus include: a div shaped like caret that fades in and out, aria-activedescendent, and canvas-based interactions with custom-rendered focus.

We've been working on a [definition of focus](https://www.w3.org/TR/WCAG22/#dfn-focus) which is fairly general, that came about through working on other issues and hasn't been linked up yet.

What I'm not sure about is the negative impact of requiring a focus indicator on things which receive keyboard focus. The game example in the google doc I didn't think had focus, but things with are active descendants should have an indicator (IMO).

**Comment by mraccess77** [comment]
It would be helpful to get clarity on whether a caret is acceptable for an input field, textarea, or element with contenteditable=true.

**Comment by alastc** [comment]
Hi @mraccess77, we have 3 'typical' examples in the understanding doc, one of which is: 

> A 1px wide vertical line (such as a blinking cursor) would not pass the size requirement, a text input would need a larger or separate focus indicator.

**Comment by alastc** [comment]
Hi @dshoukry, 

Quite a few of the comments were questions for focus-appearance which we've tried to answer in the doc and here. Could you let us know if that is satisfactory, or if there are remaining questions / comments?

Thanks,

-Alastair

**Comment by dshoukry** [comment]
Thank you! I'm gathering internal feedback and will provide an update soon.

**Comment by alastc** [comment]
Hi @dshoukry, there have been a few updates to the [SC text and understanding document](https://w3c.github.io/wcag/understanding/focus-appearance-minimum.html). We will put it out for review again soon, would it be best to close this and start a new one based on the updated version?

**Comment by dshoukry** [comment]
That works! We also have a UX Researcher who would like to present some past research on focus indicators to the group, should we try to schedule that in the new request or would it be better for me to reach out offline? Thank you!

**Comment by alastc** [comment]
Thanks @dshoukry, I'll close this one.

Could we discuss the research on email?
